Description: Added toHStore method to NamedList
Author: vir@ctm.ru
Origin: <upstream|backport|vendor|other>, <URL, required except if Author is present>
Bug: <URL to the upstream bug report if any, implies patch has been forwarded, optional>
Bug-<Vendor>: <URL to the vendor bug report if any, optional>
Forwarded: <URL|no|not-needed, useless if you have a Bug field, optional>
Applied-Upstream: <version|URL|commit, identifies patches merged upstream, optional>
Reviewed-by: <name and email of a reviewer, optional>
Last-Update: <YYYY-MM-DD, last update of the meta-information, optional>
---
This patch allows one to use '${$hstore}' substitution and pass whole Message to database functions in Postgres's HSTORE format.

Index: yate/engine/NamedList.cpp
===================================================================
--- yate.orig/engine/NamedList.cpp	2016-01-14 16:34:53.000000000 +0300
+++ yate/engine/NamedList.cpp	2016-01-15 13:49:07.000000000 +0300
@@ -246,6 +246,24 @@
     }
 }
 
+void NamedList::toHStore(String& str) const
+{
+    const ObjList *p = m_params.skipNull();
+    bool first = true;
+    for (; p; p = p->skipNext()) {
+	if(first)
+	    first = false;
+	else
+	    str << ", ";
+        const NamedString* s = static_cast<const NamedString *>(p->get());
+	str << "\"" << s->name().sqlEscape('\"') << "\" => ";
+	if(s->null())
+	    str << "NULL";
+	else
+	    str << "\"" << s->sqlEscape('\"') << "\"";
+    }
+}
+
 int NamedList::getIndex(const NamedString* param) const
 {
     if (!param)
@@ -342,6 +360,14 @@
 	    String def;
 	    String tmp = str.substr(p1+2,p2-p1-2);
 	    tmp.trimBlanks();
+	    if (tmp == YSTRING("$hstore")) {
+		tmp.clear();
+		toHStore(tmp);
+		str = str.substr(0,p1) + tmp + str.substr(p2+1);
+		p1 += tmp.length();
+		cnt++;
+		continue;
+	    }
 	    int pq = tmp.find('$');
 	    if (pq >= 0) {
 		// param is in ${<name>$<default>} format
Index: yate/yateclass.h
===================================================================
--- yate.orig/yateclass.h	2016-01-14 16:34:53.000000000 +0300
+++ yate/yateclass.h	2016-01-15 13:49:07.000000000 +0300
@@ -4800,6 +4800,12 @@
     void dump(String& str, const char* separator, char quote = 0, bool force = false) const;
 
     /**
+     * Dumps all parameters to a string in a Postgres's hstore
+     * ("key" => "value", ...) format with all required escaping.
+     */
+    void toHStore(String& str) const;
+
+    /**
      * A static empty named list
      * @return Reference to a static empty named list
      */
